### DISCUSSION TITLE
inter-clip modulation:  one clip could affect another. (x-pose, note shift etc)

### DISCUSSION THREAD

**Discussion by lobit12** [SynthstromAudible/DelugeFirmware#1292] [Ideas]
This set of ideas comes from the Eloquencer sequencer I use quite a bit..  one of the cool things it does is allow one track to affect
another.   Ive been thinking that with the new grids mode cross modulation of clips could be really interesting.

Transposition is an obvious one if you set a slower timebase on the modulator track you can get musical transpostion over several bars from a single bar or notes.   

Another great is modulating note shift from another track,  super interesting results from just a little bit of input.

Finally another cool somewhat related idea would be step advance from another clip/ track.  the way an SH101 / JX3P sequencer works so the sequence playback would only advance if it got a note from another clip.

**Comment by seangoodvibes** [comment]
> This set of ideas comes from the Eloquencer sequencer I use quite a bit..  one of the cool things it does is allow one track to affect
> 
> another.   Ive been thinking that with the new grids mode cross modulation of clips could be really interesting.
> 
> 
> 
> Transposition is an obvious one if you set a slower timebase on the modulator track you can get musical transpostion over several bars from a single bar or notes.   
> 
> 
> 
> Another great is modulating note shift from another track,  super interesting results from just a little bit of input.
> 
> 
> 
> Finally another cool somewhat related idea would be step advance from another clip/ track.  the way an SH101 / JX3P sequencer works so the sequence playback would only advance if it got a note from another clip.  
> 
> 
> 
> 
> 
> 

I believe we already have this now :) 

See: #865 and #837

Both of these features are part of our current nightly and beta builds

**Reply by lobit12** [reply]
cool.. ill investigate that, does that essentially mean we will be using midi clips to control the internal synth as well as external in this case?

**Reply by seangoodvibes** [reply]
> cool.. ill investigate that, does that essentially mean we will be using midi clips to control the internal synth as well as external in this case?

Yes that's essentially what it means

**Reply by lobit12** [reply]
is the transpose feature currently working? I couldn't figure out how to enable the midi command.

**Reply by seangoodvibes** [reply]
Yes should be working:

> MIDI Loopback - All notes and CCs from MIDI clips are sent back to Deluge, available to be learned to other clips. The behavior is as if there were a physical loopback cable, connecting Deluge's MIDI out to MIDI in. Turn on/off in Song View Sound Menu. This may be used for things like additive synthesis (one MIDI clip controls several synth instrument clips), generative melodies / polymeter rhythms (two or more MIDI clips of different lengths control the same instrument or kit clip), or macro control of sounds (have CC modulation in a separate MIDI clip that is turned on or off)..

![image](https://github.com/SynthstromAudible/DelugeFirmware/assets/138174805/8dd5079a-f8b0-4258-8f95-0e42c8cf1c03)

**Comment by duncanrmi** [comment]
love this new feature- as a long-time user of the sequentix P3 & cirklon machines, this drops nicely into my workflows.
couple of questions, though-

- what is the expected behaviour if more than one track is set to transpose other tracks?
- what exactly is the interaction with the scale setting? I feel that this could be better explained- are we simply adding to the note number arithmetically within the current scale, or is the root of the scale being changed? 
- could changes in the scale setting be made possible *using the same transposing track*? (to save creating an automation nightmare)

**Comment by duncanrmi** [comment]
been exploring this some more, & enjoying it a great deal, but feel that there's more to be had, somehow.... I would like to operate transpose on the other clips in a song from a clip that also contains an actual part, being played by midi or an internal synth/kit.
I would also like to be able to disable transpose on some clips that are still operating under the scale.
don't ask me why, I just think it would be useful.
so- UI..... how about in 'song' view, you hold down a pad with a transpose 'event' on it, & select/deselect your target tracks using the audition buttons? by default, they'd all be on (unless the clip/s are not in scale mode)

imagine a scenario.... transpose clip is transposing a bassline & two melody parts, & all of them are beholden to the scale of the song.
what I want to do is transpose the bass every four bars, but have the melody lines transposing on every second beat in those for bars, so that they're moving around a little more unpredictably.
by pressing & holding the transpose event pad, I would be able to select/deselect which other clips are affected.

it'd be useful to be able to change the scale from within the transposing clip too.

**Comment by duncanrmi** [comment]
gonna stick this here for the moment, see if anyone else is actually using this function-

internal transpose. I've got a couple of nice arpeggiated acoustic guitar parts with the deluge doing the arpeggiation, based on note clusters with probabilities set so that it gets a different choice of notes to arpeggiate each pass, & some randomness chucked in with the new arpeg features. so far so good. & a bass synth line.

so now I'm trying the internal transpose & have a couple of observations-

if you go back into any of the clips being transposed & try to edit them.... impossible. I know they're being transposed, but can we just leave the notes where they were in the untransposed version instead of having them leaping around as they get transposed?
the 'notes' doing the transposing seem to need to be a few beats early or they don't get 'processed' & nothing transposes.
the 'notes' doing the transposing seem to be issuing note-on/offs that are affecting the clips being transposed, so that long release notes are getting chopped off. I'll try that track order thing...
this is in the latest nightly, chopin.

**Reply by duncanrmi** [reply]
this behaviour is different in chromatic mode- the patterns being transposed don't leap around, & can be edited.
the processing of transpose events is timely, & the notes aren't getting cut off as much.

**Reply by duncanrmi** [reply]
fewer sums being done? definitely works better when it's not trying to keep the transpose track in key as well as calculate the consequences!

**Reply by duncanrmi** [reply]
I am finding myself greatly irritated by the way the transpose function keeps leaping around the octaves unpredictably.
_it's like it's trying to transpose itself._ 
example: I have a bass synth playing the same note every half a bar, & I am transposing this C (no transpose), A# (either down two semitones or up ten, depending how the mood takes it), G (generally goes up from the A# but sometimes goes down, so that it's still above the C), then a D# (same), but then when the C comes around again, sometimes it goes down an octave, sometimes it stays put.
all of this behaviour is even more unpredictable when you save/reload or even just stop-start a song.
having a stone-in-shoe day today, & really wishing this functionality worked like I want it to.
